### PR TITLE
expression: add setSelf for all built-in functions

### PR DESCRIPTION
**Pull request by XuHuaiyu**
We will use ScalarFunction.function.self.Eval()XXX in new type-inferer,
so it is needed to set a pointer to itself for when generating a built-in function signature.

PTAL @coocood @shenli

### CODE REVIEW

File: expression/builtin.go [hunk 1]
```diff
@@ -45,6 +45,11 @@ func newBaseBuiltinFunc(args []Expression, ctx context.Context) baseBuiltinFunc
 	}
 }
 
+func (b *baseBuiltinFunc) setSelf(f builtinFunc) builtinFunc {
+	b.self = f
```
Comment (shenli) [inline expression/builtin.go]: `b.self = b` is ok?

**Comment by shenli** [comment]
@XuHuaiyu Any update?

File: expression/builtin.go (hunk 1 above)
Comment (XuHuaiyu) [inline expression/builtin.go]: no, `baseBuiltinFunc` is not a `builtinFunc(interface`)

**Comment by shenli** [comment]
LGTM

**Pull request merged**